### DISCUSSION TITLE
Concept for storage in `swars-levels`

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#125] [General]
My idea for [swars-levels](https://github.com/swfans/swars-levels) content is now as following:

* levels will be stored in standard PC format, as this is the format any level editors will use
* maps will be stored in DAT format - based on the one we have from Pre-Alpha, and from korean version of the game
* a tool will be created which takes DAT, and uses it to generate MAD map for release and SCN file for brief screen minimap

Required new code for the DAT files:
* load DAT map file (first draft version made)
* save MAD file (not implemented)
* re-create pathfinding and traffic data which is not part of DAT but required in MAD
* close the above functionalities into a separate converter tool

Currently, maps loaded from DAT have issues; but I didn't made sure that all the structures inside are even matching my load code, so that's expected.
![synII089](https://github.com/swfans/swars/assets/11463396/ef0004a8-5051-412a-93d4-7281d28c9de4)

**Comment by mefistotelis** [comment]
Small update on DAT loading:

I re-initialized the pathfinding enough that id doesn't crash, but it does not generate a proper region map yet..
Also, traffic map is empty.

For the pre-alpha map, I also didn't mapped fields in structs which changed their sizes.
![synII101](https://github.com/swfans/swars/assets/11463396/617bf7bb-7584-43cc-910e-c933e5812b1e)

**Comment by Moburma** [comment]
A little bit off topic, but what would you like to do with optional mods? For example I've made a few things like an alternative version of the Church Dead Zone level with the obnoxious hidden exploding people toned down.
These definitely should not replace the default levels as they are mods deliberately changing the original intent, but I think it would be nice to keep them somewhere people can access if they like. The problem is of course they all use the same file names.

Do we want to support those kinds of uploads?

The other thing I might do is upload them here:
https://www.nexusmods.com/syndicatewars

**Reply by geist22** [reply]
A feature I would like to see in swars port is asset override, for example the default levels are in `LEVELS` and new levels could be put in `LEVELS_USER` which then either would be preferred over the default ones or could be enabled/disabled individually per config. That way you could use modified versions of levels without having to delete/replace the original files.

Modded files could be hosted in a repository here, or even be shipped together with the port in a disabled state to be selectively enabled by the user. It would be nice to have everything in one place without having to search around in multiple places. I don't really like the idea of outsourcing to another 3rd party, especially one of dubious integrity.

**Comment by mefistotelis** [comment]
> A little bit off topic, but what would you like to do with optional mods?

As SW Port developers, we maintain a base set of assets. If we add anything to our repos, it's then a part of the main game, not a mod.
For where to keep/share mods - this is typically decided by the modder. Nexus is  popular place for that.

> A feature I would like to see in swars port is asset override, for example the default levels are in LEVELS and new levels could be put in LEVELS_USER which then either would be preferred over the default ones or could be enabled/disabled individually per config.

The original game is built around one storage place for levels. This is an easy design, but I agree it makes modifications harder to upkeep.

What I've done in KeeperFX, and what would probably work for SW Port as well, is to make maps/levels folder a part of campaign config file.
Then original campaigns can use the original folder (though we'll likely separate them sooner or later anyway - that's what did happened with KeeperFX eventually). And any mods can have their own folders, on the mod developer discretion.

> Modded files could be hosted in a repository here, or even be shipped together with the port in a disabled state to be selectively enabled by the user.

I'm not really into "official" mods and "unofficial" mods. What the game comes with should be consistent, containing features rather than mods. Then, mods are something which either comes from the main game devs but was not accepted into the main experience, or comes from other fans/devs. Mods build something on top of the official game release, but are not a part of it.

Basically if something is accepted into the game repos, it comes with a promise that this will be maintained by the devs - if any change is made ie. in main engine, the features in official repos need to be updated for that change (like updating for changed config options, or re-balancing, or taking into consideration while making future changes to the game).

Mods are maintained by their creators - so while it should be generally avoided, a new game release can break mods. Then either the mod creator will update it, or the mod will always remain as something working for specific range of versions of the game.
